sdk/ledger-service/http-json/src/main/scala/com/digitalasset/http/json/DomainJsonDecoder.scala

```diff
@@ -194,70 +200,79 @@ class DomainJsonDecoder(
+      ciId: Option[ContractTypeId.ResolvedPkgId] = fjj.choiceInterfaceId.map(_.latestPkgId)
 
       payloadT <- either(resolveTemplateRecordType(tId).liftErr(JsonError))
-
       oIfIdArgT <- either(resolveChoiceArgType(ciId getOrElse tId, fjj.choice).liftErr(JsonError))
       (oIfaceId, argT) = oIfIdArgT
 
       payload <- either(jsValueToApiRecord(payloadT, fjj.payload))
       argument <- either(jsValueToApiValue(argT, fjj.argument))
-    } yield fjj.copy(
+
+      choiceIfaceOverride =
+        if (oIfaceId.isDefined) oIfaceId.map(i => i.map(p => Ref.PackageRef.Id(p): Ref.PackageRef))
+        else fjj.choiceInterfaceId.map(_.original)
+
+      cmd <- fjj.bitraverse(_.original.pure[ET], _.pure[ET])
+    } yield cmd.copy(
       payload = payload,
       argument = argument,
-      choiceInterfaceId = oIfaceId orElse fjj.choiceInterfaceId,
-      meta = fjj.meta,
+      choiceInterfaceId = choiceIfaceOverride,
     )
   }
 
   private[this] def jsValueToApiRecord(t: domain.LfType, v: JsValue) =
     jsValueToApiValue(t, v) flatMap mustBeApiRecord
 
   private[this] def resolveMetaTemplateIds[
-      CtId[T] <: ContractTypeId[T] with ContractTypeId.Ops[CtId, T]
+      U,
+      CtId[T] <: ContractTypeId[T] with ContractTypeId.Ops[CtId, T],
   ](
-      meta: domain.CommandMeta[CtId[Ref.PackageRef]],
+      meta: domain.CommandMeta[U with ContractTypeId.RequiredPkg],
       jwt: Jwt,
       ledgerId: LedgerApiDomain.LedgerId,
   )(implicit
       ec: ExecutionContext,
       lc: LoggingContextOf[InstanceUUID],
-  ): ET[domain.CommandMeta[ContractTypeId.ResolvedOf[CtId]]] = for {
+      overload: PackageService.ResolveContractTypeId.Overload[U, CtId],
+  ): ET[domain.CommandMeta[CtId[Ref.PackageRef]]] = for {
     // resolve as few template IDs as possible
     tpidToResolved <- {
       import scalaz.std.vector._
       val inputTpids = Foldable[domain.CommandMeta].toSet(meta)
       inputTpids.toVector
-        .traverse { ot => templateId_(ot, jwt, ledgerId) strengthL ot }
+        .traverse { ot =>
+          templateId_(ot, jwt, ledgerId).map(_.original) strengthL ot
+        }
         .map(_.toMap)
     }
   } yield meta map tpidToResolved
 
   private def templateId_[
-      CtId[T] <: ContractTypeId[T] with ContractTypeId.Ops[CtId, T]
+      U,
+      CtId[T] <: ContractTypeId[T] with ContractTypeId.Ops[CtId, T],
   ](
-      id: CtId[Ref.PackageRef],
+      id: U with ContractTypeId.RequiredPkg,
       jwt: Jwt,
       ledgerId: LedgerApiDomain.LedgerId,
   )(implicit
       ec: ExecutionContext,
       lc: LoggingContextOf[InstanceUUID],
-  ): ET[domain.ContractTypeId.ResolvedOf[CtId]] =
+      overload: PackageService.ResolveContractTypeId.Overload[U, CtId],
+  ): ET[domain.ContractTypeRef[CtId]] =
     eitherT(
       resolveContractTypeId(jwt, ledgerId)(id)
         .map(
           _.toOption.flatten
-            .map(_.original.asInstanceOf[ContractTypeId.ResolvedOf[CtId]])
```

Comment:
Previously `templateId_` would internally determine the `ContractTypeRef` for the input id, and then would always return the `original` field, which was a resolved version of it.

The change is to return the `ContractTypeRef` itself here, and let the caller determined where they want to use the `original` or `latestPkgId` field. The `latestPkgId` field must be used for all type lookups, but the `original` may be used for representing the resolved templateId in downstream types.